File: keeper/src/main/java/com/github/abhijitpparate/keeper/screen/note/presenter/NotePresenter.java
package com.github.abhijitpparate.keeper.screen.note.presenter;


import android.net.Uri;
import android.util.Log;

import com.github.abhijitpparate.keeper.data.auth.AuthSource;
import com.github.abhijitpparate.keeper.data.auth.User;
import com.github.abhijitpparate.keeper.data.database.DatabaseSource;
import com.github.abhijitpparate.keeper.data.database.Note;
import com.github.abhijitpparate.keeper.data.storage.File;
import com.github.abhijitpparate.keeper.data.storage.StorageInjector;
import com.github.abhijitpparate.keeper.data.storage.StorageSource;
import com.github.abhijitpparate.keeper.scheduler.SchedulerProvider;
import com.github.abhijitpparate.keeper.utils.Utils;
import com.github.abhijitpparate.keeper.R;
import com.github.abhijitpparate.keeper.data.auth.AuthInjector;
import com.github.abhijitpparate.keeper.data.database.DatabaseInjector;
import com.github.abhijitpparate.keeper.scheduler.SchedulerInjector;
import com.google.android.gms.location.places.Place;

import java.util.HashMap;
import java.util.Map;

import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableCompletableObserver;
import io.reactivex.observers.DisposableMaybeObserver;

public class NotePresenter implements NoteContract.Presenter {

    public static final String TAG = "NotePresenter";

    public static Map<String, Integer> colorMap = new HashMap<>();
    static {
        colorMap.put(Note.NoteColor.DEFAULT,   R.color.colorBackground);
        colorMap.put(Note.NoteColor.WHITE,     android.R.color.white);
        colorMap.put(Note.NoteColor.RED,       android.R.color.holo_red_light);
        colorMap.put(Note.NoteColor.GREEN,     android.R.color.holo_green_light);
        colorMap.put(Note.NoteColor.YELLOW,    android.R.color.holo_orange_light);
        colorMap.put(Note.NoteColor.BLUE,      android.R.color.holo_blue_bright);
        colorMap.put(Note.NoteColor.ORANGE,    android.R.color.holo_orange_dark);
    }

    private SchedulerProvider schedulerProvider;
    private CompositeDisposable disposable;

    private AuthSource authSource;
    private DatabaseSource databaseSource;
    private StorageSource storageSource;

    private User currentUser;
    private Note currentNote;
    private Uri tempFileUri;

    private NoteContract.View view;

    private boolean isOptionsOpen = false;

    public NotePresenter(NoteContract.View view) {
        this.view = view;

        this.authSource = AuthInjector.getAuthSource();
        this.databaseSource = DatabaseInjector.getDatabaseSource();
        this.storageSource = StorageInjector.getStorageSource();

        this.schedulerProvider = SchedulerInjector.getScheduler();
        this.disposable = new CompositeDisposable();
        this.currentNote = Note.newNote();

        view.setPresenter(this);
    }

    private Note assembleNote() {
//        if (currentNote == null) currentNote = Note.newNote();

//        if (!currentNote.getNoteId().equals(view.getNoteId())) {
//            currentNote = Note.newNote();
//        }

        currentNote.setTitle(view.getNoteTitle());
        currentNote.setBody(view.getNoteBody());
        currentNote.setChecklist(view.getCheckList());
//        currentNote.setColor(view.getNoteColor());
//        currentNote.setPlace(view.getNotePlace());
//        currentNote.setFile(view.getFile());
        return currentNote;
    }

    @Override
    public void onSaveClick() {
        Note note = assembleNote();
        if (tempFileUri != null){
            saveFile(note);
        } else
            saveNote(note);
    }

    public void saveFile(final Note note){
        disposable.add(
                storageSource
                        .createFile(currentUser.getUid(), note.getFile().getName(), note.getFile().getType(), tempFileUri)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(new DisposableMaybeObserver<File>(){

                            @Override
                            public void onSuccess(File file) {
                                Log.d(TAG, "onSuccess: ");
                                note.setFile(file);
                                saveNote(note);
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.d(TAG, "onError: ");
                                view.makeToast(e.getMessage());
                            }

                            @Override
                            public void onComplete() {
                                Log.d(TAG, "onComplete: ");
                                tempFileUri = null;
                                view.makeToast("onComplete");
                            }
                        })
        );
    }

    public void saveNote(Note note){
        Log.d(TAG, "saveNote: ");
        disposable.add(
                databaseSource
                        .createOrUpdateNote(currentUser.getUid(), note)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                new DisposableCompletableObserver() {
                                    @Override
                                    public void onComplete() {
                                        view.showHomeScreen();
                                    }

                                    @Override
                                    public void onError(Throwable e) {
                                        view.makeToast(e.getMessage());
                                    }
                                }
                        )
        );
    }

    @Override
    public void loadNote(final String noteId) {
//        Log.d(TAG, "loadNote: ");
        view.showProgressbar(true);
        disposable.add(
                databaseSource
                        .getNoteFromId(currentUser.getUid(), noteId)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                new DisposableMaybeObserver<Note>() {

                                    @Override
                                    public void onComplete() {

                                    }

                                    @Override
                                    public void onSuccess(@NonNull Note note) {
                                        NotePresenter.this.currentNote = note;
//                                        view.setNote(note);

                                        view.setNoteTitle(note.getTitle());

                                        if (note.getBody() != null && !note.getBody().isEmpty()){
                                            view.setNoteBody(note.getBody());
                                        } else if (note.getChecklist() != null && !note.getChecklist().equals("[]")){
                                            view.setNoteChecklist(Utils.parseNoteChecklist(note.getChecklist()));
                                            view.switchToChecklist();
                                        }

                                        view.setNoteColor(Utils.getNoteColor(note.getColor()), note.getColor() != null ? note.getColor() : Note.NoteColor.DEFAULT);

                                        if (note.getPlace() != null && note.getPlace().getLocation() != null) {
                                            view.setNotePlace(note.getPlace());
                                        }

                                        if (note.getFile() != null) {
                                            view.setNoteFile(note.getFile());
                                        }

                                        view.showProgressbar(false);
                                    }

                                    @Override
                                    public void onError(Throwable e) {
                                        view.makeToast(e.getMessage());
                                    }
                                }
                        )
        );
    }

    @Override
    public void subscribe() {
        getUserData();
    }

    private void getUserData() {
//        Log.d(TAG, "getUserData: ");
        disposable.add(
                authSource
                        .getUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(new DisposableMaybeObserver<User>() {
                            @Override
                            public void onSuccess(@NonNull User user) {
                                currentUser = user;
//                                Log.d(TAG, "onSuccess: ");
                                view.loadNoteIfAvailable();
                            }

                            @Override
                            public void onError(@NonNull Throwable e) {
//                                Log.d(TAG, "onError: ");
                            }

                            @Override
                            public void onComplete() {
//                                Log.d(TAG, "onComplete: ");
                            }
                        })
        );
    }

    @Override
    public void unsubscribe() {
        disposable.clear();
    }

    @Override
    public void onChecklistClick(boolean isChecked) {
        if (isChecked) view.switchToChecklist();
        else view.switchToText();
    }

    @Override
    public void onDrawingClick() {

    }

    @Override
    public void onLocationClick() {
        view.showPlacePicker();
    }

    @Override
    public void onAudioClick() {

    }

    @Override
    public void onVideoClick() {

    }

    @Override
    public void onImageClick() {
        view.showFilePicker();
    }

    @Override
    public void onOptionsClick() {
        toggleOptionsPanel();
    }

    private void toggleOptionsPanel(){
        if (isOptionsOpen){
            view.hideOptionsPanel();
        } else {
            view.showOptionsPanel();
        }
        isOptionsOpen = !isOptionsOpen;
    }

    @Override
    public void onDeleteClick() {
        Log.d(TAG, "onDeleteClick: ");
        toggleOptionsPanel();
    }

    @Override
    public void onSendClick() {
        Log.d(TAG, "onSendClick: ");
        toggleOptionsPanel();
    }

    @Override
    public void onDuplicateClick() {
        Log.d(TAG, "onDuplicateClick: ");
        toggleOptionsPanel();
    }

    @Override
    public void onLabelClick() {
        Log.d(TAG, "onLabelClick: ");
        toggleOptionsPanel();
    }

    @Override
    public void onColorSelected(NoteContract.NoteColor color) {
        Log.d(TAG, "onColorSelected: " + color);
        switch (color){
            case WHITE:
                view.setNoteColor(android.R.color.white, Note.NoteColor.WHITE);
                break;
            case RED:
                view.setNoteColor(android.R.color.holo_red_light, Note.NoteColor.RED);
                break;
            case GREEN:
                view.setNoteColor(android.R.color.holo_green_light, Note.NoteColor.GREEN);
                break;
            case YELLOW:
                view.setNoteColor(android.R.color.holo_orange_light, Note.NoteColor.YELLOW);
                break;
            case BLUE:
                view.setNoteColor(android.R.color.holo_blue_bright, Note.NoteColor.BLUE);
                break;
            case ORANGE:
                view.setNoteColor(android.R.color.holo_orange_dark, Note.NoteColor.ORANGE);
                break;
            default:
                view.setNoteColor(R.color.colorBackground, Note.NoteColor.DEFAULT);
                break;
        }
    }

    @Override
    public void onPlaceSelected(Place place) {
        Note.Place p = new Note.Place();
        p.setName(place.getName().toString());
        p.setLocation(new Note.Place.LatLng(place.getLatLng().latitude, place.getLatLng().longitude));
        view.setNotePlace(p);
    }

    @Override
    public void onFileSelected(File file, Uri uri) {
        this.tempFileUri = uri;
        this.currentNote.setFile(file);
    }
}

File: keeper/src/main/java/com/github/abhijitpparate/keeper/screen/note/presenter/NotePresenter.kt
package com.github.abhijitpparate.keeper.screen.note.presenter


import android.util.Log

import com.github.abhijitpparate.keeper.R
import com.github.abhijitpparate.keeper.data.auth.AuthInjector
import com.github.abhijitpparate.keeper.data.auth.AuthSource
import com.github.abhijitpparate.keeper.data.auth.User
import com.github.abhijitpparate.keeper.data.database.DatabaseInjector
import com.github.abhijitpparate.keeper.data.database.DatabaseSource
import com.github.abhijitpparate.keeper.data.database.Note
import com.github.abhijitpparate.keeper.scheduler.SchedulerInjector
import com.github.abhijitpparate.keeper.scheduler.SchedulerProvider
import com.github.abhijitpparate.keeper.utils.Utils

import java.util.HashMap

import io.reactivex.annotations.NonNull
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableCompletableObserver
import io.reactivex.observers.DisposableMaybeObserver

import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.BLUE
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.DEFAULT
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.GREEN
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.ORANGE
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.RED
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.WHITE
import com.github.abhijitpparate.keeper.data.database.Note.NoteColor.YELLOW
import com.github.abhijitpparate.keeper.utils.Utils.getNoteColor
import com.google.android.gms.location.places.Place

class NotePresenter(private var view: NoteContract.View) : NoteContract.Presenter {

    private var schedulerProvider: SchedulerProvider = SchedulerInjector.scheduler
    private var disposable: CompositeDisposable = CompositeDisposable()
    private var authSource: AuthSource = AuthInjector.authSource
    private var databaseSource: DatabaseSource = DatabaseInjector.databaseSource

    lateinit var currentUser: User
    private var currentNote: Note? = null

    private var isOptionsOpen = false

    init {
        view.setPresenter(this)
    }

    override fun onSaveClick() {
        assembleNote()
        disposable.add(
                databaseSource
                        .createOrUpdateNote(currentUser.uid.toString(), currentNote!!)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                object : DisposableCompletableObserver() {
                                    override fun onComplete() {
                                        view.showHomeScreen()
                                    }

                                    override fun onError(e: Throwable) {
                                        view.makeToast(e.message.toString())
                                    }
                                }
                        )
        )
    }

    private fun assembleNote(): Note {
        if (currentNote == null) currentNote = Note()

        if (!currentNote?.noteId.equals(view.getNoteUuid())) {
            currentNote = Note()
        }

        currentNote!!.title = view.getNoteTitle()
        currentNote!!.body = view.getNoteBody()
        currentNote!!.checklist = view.getNoteCheckList()
        currentNote!!.color = view.getNoteColor()
        currentNote!!.place = view.getNotePlace()
        return currentNote as Note
    }

    override fun loadNote(noteId: String) {
//        Log.d(TAG, "loadNote: ")
        view.showProgressbar(true)
        disposable.add(
                databaseSource
                        .getNoteFromId(currentUser.uid.toString(), noteId)
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(
                                object : DisposableMaybeObserver<Note>() {

                                    override fun onComplete() {

                                    }

                                    override fun onSuccess(@NonNull note: Note) {
                                        currentNote = note
                                        view.setNote(note)
                                        view.setNoteTitle(note.title)
                                        if (note.body != null && note.body?.isNotEmpty() as Boolean) {
                                            view.setNoteBody(note.body)
                                        } else if (note.checklist != null && note.checklist != "[]") {
                                            view.setNoteChecklist(Utils.parseNoteChecklist(note.checklist!!))
                                            view.switchToChecklist()
                                        }
                                        note.color?.let {
                                            view.setNoteColor(getNoteColor(note.color), it)
                                        }

                                        note.place?.let {
                                            Log.d(TAG, note.place?.name)
                                            view.setNoteLocation(note.place as Note.Place)
                                        }
                                        view.showProgressbar(false)
                                    }

                                    override fun onError(e: Throwable) {
                                        view.makeToast(e.message.toString())
                                    }
                                }
                        )
        )
    }

    override fun subscribe() {
        getUserData()
    }

    private fun getUserData() {
        //        Log.d(TAG, "getUserData: ");
        disposable.add(
                authSource
                        .retrieveUser()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribeWith(object : DisposableMaybeObserver<User>() {
                            override fun onSuccess(@NonNull user: User) {
                                currentUser = user
                                view.loadNoteIfAvailable()
                            }

                            override fun onError(@NonNull e: Throwable) {

                            }

                            override fun onComplete() {
                            }
                        })
        )
    }

    override fun unsubscribe() {
        disposable.clear()
    }

    override fun onChecklistClick(isChecked: Boolean) {
        if (isChecked)
            view.switchToChecklist()
        else
            view.switchToText()
    }

    override fun onDrawingClick() {

    }

    override fun onLocationClick() {
        view.showLocationPicker()
    }

    override fun onAudioClick() {

    }

    override fun onVideoClick() {

    }

    override fun onImageClick() {

    }

    override fun onOptionsClick() {
        toggleOptionsPanel()
    }

    fun toggleOptionsPanel() {
        if (isOptionsOpen) {
            view.hideOptionsPanel()
        } else {
            view.showOptionsPanel()
        }
        isOptionsOpen = !isOptionsOpen
    }

    override fun onDeleteClick() {
        Log.d(TAG, "onDeleteClick: ")
        toggleOptionsPanel()
    }

    override fun onSendClick() {
        Log.d(TAG, "onSendClick: ")
        toggleOptionsPanel()
    }

    override fun onDuplicateClick() {
        Log.d(TAG, "onDuplicateClick: ")
        toggleOptionsPanel()
    }

    override fun onLabelClick() {
        Log.d(TAG, "onLabelClick: ")
        toggleOptionsPanel()
    }

    override fun onColorSelected(color: NoteContract.NoteColor) {
        Log.d(TAG, "onColorSelected: " + color)
        when (color) {
            NoteContract.NoteColor.WHITE -> view.setNoteColor(android.R.color.white, WHITE)
            NoteContract.NoteColor.RED -> view.setNoteColor(R.color.noteColorRed, RED)
            NoteContract.NoteColor.GREEN -> view.setNoteColor(R.color.noteColorGreen, GREEN)
            NoteContract.NoteColor.YELLOW -> view.setNoteColor(R.color.noteColorYellow, YELLOW)
            NoteContract.NoteColor.BLUE -> view.setNoteColor(R.color.noteColorBlue, BLUE)
            NoteContract.NoteColor.ORANGE -> view.setNoteColor(R.color.noteColorOrange, ORANGE)
            else -> view.setNoteColor(R.color.colorBackground, DEFAULT)
        }
    }

    override fun onLocationSelected(location: Place) {
        val place:Note.Place = Note.Place()
        place.location?.lat = location.latLng.latitude
        place.location?.lng = location.latLng.longitude
        place.name = location.name.toString()
        currentNote?.place = place
        view.setNoteLocation(place)
    }

    companion object {

        val TAG = "NotePresenter"

        @JvmStatic var colorMap: MutableMap<String, Int> = HashMap()

        init {
            colorMap.put(DEFAULT,   R.color.colorBackground)
            colorMap.put(WHITE,     android.R.color.white)
            colorMap.put(RED,       R.color.noteColorRed)
            colorMap.put(GREEN,     R.color.noteColorGreen)
            colorMap.put(YELLOW,    R.color.noteColorYellow)
            colorMap.put(BLUE,      R.color.noteColorBlue)
            colorMap.put(ORANGE,    R.color.noteColorOrange)
        }
    }
}
